compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -442,21 +447,37 @@ static BOOL_TYS: TyS<'static> = TyS {
     kind: ty::Bool,
     flags: TypeFlags::empty(),
     outer_exclusive_binder: DebruijnIndex::from_usize(0),
+    stable_hash: Fingerprint::ZERO,
 };
 
 impl<'a, 'tcx> HashStable<StableHashingContext<'a>> for Ty<'tcx> {
     fn hash_stable(&self, hcx: &mut StableHashingContext<'a>, hasher: &mut StableHasher) {
         let TyS {
-            ref kind,
+            kind,
 
             // The other fields just provide fast access to information that is
             // also contained in `kind`, so no need to hash them.
             flags: _,
 
             outer_exclusive_binder: _,
+
+            stable_hash,
         } = self.0.0;
 
-        kind.hash_stable(hcx, hasher);
+        if *stable_hash == Fingerprint::ZERO {
+            // No cached hash available. This can only mean that incremental is disabled.
+            // We don't cache stable hashes in non-incremental mode, because they are used
+            // so rarely that the performance actually suffers.
+
+            let stable_hash: Fingerprint = {
+                let mut hasher = StableHasher::new();
+                kind.hash_stable(hcx, &mut hasher);
+                hasher.finish()
+            };
+            stable_hash.hash_stable(hcx, hasher);
+        } else {
+            stable_hash.hash_stable(hcx, hasher);
+        }
```

Comment:
It's probably an issue if stable hashes differ between incremental and non-incremental mode. I can imagine that, for example, it leads to linker errors if an upstream crate is compiled in non-incremental mode and then a down-stream crate is compiled in incremental mode, because the downstream crate computes a different symbol hash.

Caching the hash inside of a std::lazy::once (or something similar) might be worth looking into.

Reply:
I just saw that you're hashing the same thing in both cases, so nevermind.